Refactor hot methods

DomGarguilo opened the pull request:
In #1099 there are some lists of methods that have been reported as "hot method too big".

This PR refactors some of those methods to make them smaller. After these changes, the following methods are no longer reported as "hot method too big":
* `org.apache.accumulo.core.file.rfile.RelativeKey::<init>`
* `org.apache.accumulo.core.file.rfile.RelativeKey::readFields`
* `org.apache.accumulo.server.fs.FileManager$ScanFileManager::openFiles`

The following have been refactored to reduce size, but are still being reported:
* `org.apache.accumulo.core.iteratorsImpl.system.LocalityGroupIterator::_seek`
* `org.apache.accumulo.tserver.scan.LookupTask::run`

DomGarguilo commented:
I have marked this as WIP since there is still work to be done here:
* Look into refactoring more methods appearing in the list of methods in #1099 
* Rename new methods. I am not sure what to name the new methods so suggestions are very welcome.
  *  Namely, the new methods in `RelativeKey`: `extracted()` and `getBytes()`. Both of which I have left as thier IDE generated names

EdColeman commented:
Looking at the method in 'Key' the method that performs the equality check on a byte array  `private static boolean isEqual(byte[] a1, byte[] a2)` is already optimized for Accumulo row comparisons, checking the last few bytes first and does not seem to need additional optimization.

DomGarguilo commented:
> Looking at the method in 'Key' the method that performs the equality check on a byte array `private static boolean isEqual(byte[] a1, byte[] a2)` is already optimized for Accumulo row comparisons, checking the last few bytes first and does not seem to need additional optimization.

Is this in relation to #2812? If so, do you think its worth investigating rearranging some of the comparisons made as mentioned in your comments [here](https://github.com/apache/accumulo/issues/2812#issuecomment-1188111321), or do you think #2812 should be closed?

DomGarguilo commented:
A few more methods were found ([details](https://github.com/apache/accumulo/issues/1099#issuecomment-1198342829)). I'll look into those now, before this is merged.

DomGarguilo commented:
> A few more methods were found ([details](https://github.com/apache/accumulo/issues/1099#issuecomment-1198342829)). I'll look into those now, before this is merged.

Never mind. Took a look and I don't think those methods will be changed in this PR. This PR is ready for review.

DomGarguilo commented:
I have refactored `Key.equals()` in a way that makes it not show up in the hot methods list any more. Here is the method after refactoring:
```java
  public boolean equals(Key other, PartialKey part) {
    boolean result = isEqual(row, other.row);
    if (part == ROW)
      return result;
    result &= isEqual(colFamily, other.colFamily);
    if (part == ROW_COLFAM)
      return result;
    result &= isEqual(colQualifier, other.colQualifier);
    if (part == ROW_COLFAM_COLQUAL)
      return result;
    result &= isEqual(colVisibility, other.colVisibility);
    if (part == ROW_COLFAM_COLQUAL_COLVIS)
      return result;
    result &= (timestamp == other.timestamp);
    if (part == ROW_COLFAM_COLQUAL_COLVIS_TIME)
      return result;
    result &= (deleted == other.deleted);
    if (part == ROW_COLFAM_COLQUAL_COLVIS_TIME_DEL)
      return result;

    throw new IllegalArgumentException("Unrecognized partial key specification " + part);
  }
```
It will fall through and compare each `PartialKey` until it reaches `part` and returns. Here is the current code to compare:
https://github.com/apache/accumulo/blob/388d3ffe11b3eaffe5a0022907d10e18bef73253/core/src/main/java/org/apache/accumulo/core/data/Key.java#L977-L1002

While this refactored method is smaller and no longer "hot", it should probably be determined if this has any significant performance hits which is what is being invstigated in #2812

milleruntime commented:
Thanks for taking a look at these methods. I will take a look at this today.

milleruntime commented on core/src/main/java/org/apache/accumulo/core/iteratorsImpl/system/LocalityGroupIterator.java:
```diff
@@ -163,26 +164,14 @@ static final Collection<LocalityGroup> _seek(HeapIterator hiter, LocalityGroupCo
-      }
+      groups = inclusive ? List.of() : lgContext.groups;
     } else {
       groups = new HashSet<>();
```
You might be able to create a method for the work done in this else. Something like getGroups().

milleruntime replied:
Or create a method for all the logic for "determining the set of groups to use".

DomGarguilo replied:
Extracted all of the logic into its own method in [4b51e91](https://github.com/apache/accumulo/pull/2811/commits/4b51e91e20510be1c5913e68e0d8fbba2be0d361)

milleruntime reviewed (commented):
I am not sure if it makes a difference with the compiler but we have a bad habit of jamming static classes within another class. A lot of these could be pulled out into their own class.

milleruntime commented on server/tserver/src/main/java/org/apache/accumulo/tserver/TabletClientHandler.java:
```diff
@@ -1202,33 +1202,25 @@ public void flush(TInfo tinfo, TCredentials credentials, String lock, String tab
-    }
-
-    Long flushID = null;
+    final Long[] flushID = {null};
```
I think this workaround to make the object final could give you a ConcurrentModificationException or some type of concurrency problems. They made it a compile time error for a reason, this answer here gives a good explanation https://stackoverflow.com/questions/34865383/variable-used-in-lambda-expression-should-be-final-or-effectively-final

I am not sure how the equivalent logic could be accomplished with a stream. It may be better to use a traditional for loop.

DomGarguilo replied:
Good catch. Should be fixed by [09986b2](https://github.com/apache/accumulo/pull/2811/commits/09986b2f730c2e5cd0667de5bb01a0f04615955b).

DomGarguilo commented:
In 89a2f32 I added [the equals method described above](https://github.com/apache/accumulo/pull/2811#issuecomment-1198615354). This updated method does not appear on the hot methods list.

milleruntime commented:
@DomGarguilo is this PR ready to merge? If so, I can take another close look at it.

DomGarguilo commented:
> @DomGarguilo is this PR ready to merge? If so, I can take another close look at it.

Yes, I think I am done making changes here and was just waiting for someone to verify these changes.

DomGarguilo commented on core/src/main/java/org/apache/accumulo/core/data/Key.java:
```diff
@@ -975,30 +981,26 @@ public void write(DataOutput out) throws IOException {
+    if (part == ROW_COLFAM_COLQUAL_COLVIS_TIME)
+      return result;
+    result &= (deleted == other.deleted);
+    if (part == ROW_COLFAM_COLQUAL_COLVIS_TIME_DEL)
```
Good catch. I'll revert this to how it was.

DomGarguilo replied:
@keith-turner this has been reverted in 5238f81

keith-turner commented on server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/CompactableImpl.java:
```diff
@@ -396,80 +396,89 @@ Set<StoredTabletFile> getCandidates(Set<StoredTabletFile> currFiles, CompactionK
+            Set<StoredTabletFile> candidates = Sets.difference(selectedFiles, allCompactingFiles);
+            Preconditions.checkState(currFiles.containsAll(candidates),
+                "selected files not in all files %s %s", candidates, currFiles);
+            return Collections.unmodifiableSet(candidates);
```
This code is called in a sync block in which `selectedFiles` and `allCompactingFiles` will not be changing.  The functions `Sets.difference()` and `Collections.unmodifiableSet` may keep references to the passed in sets but will not copy.  So this function may return a set that internally references   `selectedFiles` and `allCompactingFiles` which may be changing.  Its therefore important to copy as the original code did.

The following change will copy.  I have no idea if its less or more efficient than the original code, but I think `Sets.difference()` plus `Set.copyOf()` is more readable than the original code so I would go with it. 

```suggestion
            //must create a copy because the sets passed to Sets.difference could change after this method returns
            return Set.copyOf(candidates);
```

I looked around at the other changes and I think all of those copy, but it would be good if you could double check for any other cases like this in the changes in this file.

DomGarguilo replied:
Addressed in 7b5a7d9. I added `Set.copyOf()` to one other spot so take a look and see what you think.

keith-turner replied:
> I added Set.copyOf() to one other spot so take a look and see what you think.

yeah that needed the same change, I missed that.

DomGarguilo commented on server/tserver/src/main/java/org/apache/accumulo/tserver/scan/LookupTask.java:
```diff
@@ -95,9 +95,14 @@ public void run() {
-        List<Range> ranges = entry.getValue();
+        final KeyExtent extent;
+        final List<Range> ranges;
+        {
+          final Entry<KeyExtent,List<Range>> entry = iter.next();
+          extent = entry.getKey();
+          ranges = entry.getValue();
+        }
```
> Was this little code block created just to narrowly scope `entry`?

Yes. Maybe its not worth doing as it might cause confusion? What do you think?

keith-turner replied:
> Yes. Maybe its not worth doing as it might cause confusion? What do you think?

When I first saw it I had no idea why it was there.  I looked at it for a bit and guessed why it might be there. I think its an interesting technique and it reminds me of how in golang a function can return zero or more things.  Where in java a functions can only return zero or one things.   I really like how golang works.   To bad we can't write the following in java

```java
(entry,ranges) = iter.next();
```

Anyway, I am not completely sure if it should say or go, but I am leaning towards removing it.   One thing that helped me think through this is asking would this be a good practice throughout the code?   It seems like this technique could add a lot of noise that may make the code harder to read if used frequently, so that's a reason not to use it elsewhere and here.  The tight scoping could help avoid some bugs in larger methods, but maybe if we have a large method we just need to make it smaller like you have done for methods in this PR.

DomGarguilo replied:
Addressed in c86dac5

DomGarguilo commented on core/src/main/java/org/apache/accumulo/core/file/rfile/RelativeKey.java:
```diff
@@ -135,6 +118,17 @@ else if (cvCommonPrefixLen > 1)
+      byte commonPrefix) {
+    int commonPrefixLen = getCommonPrefix(prevKeyScratch, keyScratch);
+    if (commonPrefixLen == -1) {
+      fieldsSame |= rowSame;
```
Addressed in c86dac5

DomGarguilo commented on core/src/main/java/org/apache/accumulo/core/iteratorsImpl/system/LocalityGroupIterator.java:
```diff
@@ -207,35 +208,33 @@ static final Collection<LocalityGroup> _seek(HeapIterator hiter, LocalityGroupCo
+      if (columnFamilies instanceof Set<?>) {
+        cfSet = (Set<ByteSequence>) columnFamilies;
+      } else {
+        cfSet = new HashSet<>(columnFamilies);
```
Addressed in c86dac5

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/file/rfile/RelativeKey.java:
```diff
@@ -174,40 +168,13 @@ public void readFields(DataInput in) throws IOException {
-      cv = read(in);
-    }
+    row = getData(in, ROW_SAME, ROW_COMMON_PREFIX, () -> prevKey.getRowData());
+    cf = getData(in, CF_SAME, CF_COMMON_PREFIX, () -> prevKey.getColumnFamilyData());
+    cq = getData(in, CQ_SAME, CQ_COMMON_PREFIX, () -> prevKey.getColumnQualifierData());
+    cv = getData(in, CV_SAME, CV_COMMON_PREFIX, () -> prevKey.getColumnVisibilityData());
```
I love the use of lambdas, but I think these can just use the form like `prevKey::getColumnFamilyData`, right?

DomGarguilo replied:
Yea looks like it. Nice catch. Included in 59e8357

DomGarguilo replied:
For some reason this is causing some tests to fail so I am going to revert the changes in 59e8357

DomGarguilo replied:
Reverted in [9a042ab](https://github.com/apache/accumulo/pull/2811/commits/9a042ab29a668c1e8a8d036ebf724f95ceedd8c6)

ctubbsii replied:
Can you provide more details on the test failures this causes? It seems odd that it would cause test failures, since this should just be syntactic sugar and not a functional change.

ctubbsii replied:
Oh, nevermind. I think I see the test failures. It looks like you can't use `variable::method` on a null variable, but `() -> variable.method()` works fine, so long as it doesn't actually get executed. That's a bit weird, but I'm okay leaving this as-is if it works around a test failure.

DomGarguilo replied:
The test errors can be seen [here](https://github.com/apache/accumulo/actions/runs/3152941923/jobs/5128865689#step:6:2393)

here are the failing tests:
```
[ERROR] Errors: 
[ERROR]   RFileClientTest.testCache:523->lambda$testCache$0:525 » NullPointer
[ERROR]   BlockIndexTest.test1:91 » NullPointer
[ERROR]   BlockIndexTest.testSame:163 » NullPointer
[ERROR]   RFileTest.test11:1213 » NullPointer
[ERROR]   RFileTest.test17:1409 » NullPointer
[ERROR]   RFileTest.test18:1542->t18Verify:1494 » NullPointer
[ERROR]   RFileTest.test3:472 » NullPointer
[ERROR]   RFileTest.test8:758 » NullPointer
[ERROR]   RFileTest.testEncRFile11:1842->test11:1213 » NullPointer
[ERROR]   RFileTest.testEncRFile17:1876->test17:1409 » NullPointer
[ERROR]   RFileTest.testEncRFile18:1882->test18:1542->t18Verify:1494 » NullPointer
[ERROR]   RFileTest.testEncRFile3:1786->test3:472 » NullPointer
[ERROR]   RFileTest.testEncRFile8:1821->test8:758 » NullPointer
[ERROR]   RFileTest.testEncSample:2195->testSampleLG:2174->checkSample:2006 » NullPointer
[ERROR]   RFileTest.testEncryptedRFiles:1898->test3:472 » NullPointer
[ERROR]   RFileTest.testOldVersions:1684->runVersionTest:1727 » NullPointer
[ERROR]   RFileTest.testOldVersionsWithCrypto:1693->runVersionTest:1727 » NullPointer
[ERROR]   RFileTest.testReseekUnconsumed:1634 » NullPointer
[ERROR]   RFileTest.testSampleLG:2174->checkSample:2006 » NullPointe
```
Locally i kept stepping back through commits to determine that commit was causing these failures. I also think its weird that this causes failures but it seems like it is.

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/spi/balancer/SimpleLoadBalancer.java:
```diff
@@ -297,10 +275,7 @@ List<TabletMigration> move(ServerCounts tooMuch, ServerCounts tooLittle, int cou
-      if (tooLittleCount == null) {
-        tooLittleCount = 0;
-      }
+      Integer tooLittleCount = tooLittleMap.getOrDefault(table, 0);
```
I briefly looked, and didn't see an Integer.ZERO. Such a thing would be useful here, if it existed, so we don't have to keep boxing `0`. But, I'm sure this is fine. The JRE can optimize stuff like this if it wanted to.

ctubbsii reviewed (commented):
I didn't finish looking at all of this. I realized it may be easier to try to fix one thing at a time, rather than try to fix all hot methods in a single big PR. Smaller, more narrowly scoped PRs get reviewed and merged faster, and you can build on them incrementally with lower risk. And, there's less drift and merge conflicts when smaller PRs get merged faster.
